src/main/java/me/moodcat/soundcloud/SoundCloudExtract.java

```diff
@@ -0,0 +1,138 @@
+     */
+    public SoundCloudTrack extract(String artist, String permalink) throws IOException {
+        String infoUrl = resolveUrl(artist, permalink);
+
+        return parseInfoJson(infoUrl);
+    }
+
+    /**
+     * Retrieve a SoundCloudTrack given a SoundCloud URL.
+     *
+     * @param soundCloudUrl
+     *            the give SoundCloud URL
+     * @return the parsed {@link SoundCloudTrack}
+     * @throws SoundCloudException
+     *             if the URL is malformed
+     */
+    public SoundCloudTrack extract(String soundCloudUrl) throws IOException, SoundCloudException {
+        Matcher matcher = songPattern.matcher(soundCloudUrl);
+
+        if (matcher.find()) {
+            String permalink = matcher.group("permalink");
+            String artist = matcher.group("artist");
+
+            return extract(artist, permalink);
+        }
+
+        throw new SoundCloudException("Wrong URL supplied");
+    }
+
+    /**
+     * Parse the stream URL of a not-downloadable SoundCloud track. This is done
+     * by downloading and parsing a JSON response from the SoundCloud API.
+     *
+     * @param song
+     *            the song to parse the stream URL of
+     * @return the URL of the stream
+     * @throws SoundCloudException
+     *             if the stream could not be parsed
+     * @throws IOException
+     *             if the mediaURL is malformed or if the JSON page could not be
+     *             downloaded.
+     */
+    protected String parseStreamUrl(SoundCloudTrack song) throws SoundCloudException, IOException {
+        String streamJsonUrl = "http://api.soundcloud.com/i1/tracks/" + song.getId()
+                + "/streams?client_id=" + CLIENT_ID + "&secret_token=None";
+
+        String jsonPage = IOUtils.toString(new URL(streamJsonUrl));
+
+        JSONObject root = new JSONObject(jsonPage);
+
+        if (root.has("http_mp3_128_url")) {
+            return root.getString("http_mp3_128_url");
+        } else if (root.has("hls_mp3_128_url")) {
+            return root.getString("hls_mp3_128_url");
+        } else if (root.has("preview_mp3_128_url")) {
+            return root.getString("preview_mp3_128_url");
+        }
+
+        throw new SoundCloudException("No stream URL found");
+    }
+
+    /**
+     * Resolve the URL of the SoundCloud track given the artist and title.
+     *
+     * @param artist
+     *            the artist
+     * @param title
+     *            the title
+     * @return the resolved URL
+     */
+    protected String resolveUrl(String artist, String title) {
+        String url = null;
+
+        try {
+            url = "https://soundcloud.com/" + URLEncoder.encode(artist, "UTF-8") + "/"
+                    + URLEncoder.encode(title, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+        }
+
+        return "https://api.soundcloud.com/resolve.json?url=" + url + "&client_id=" + CLIENT_ID;
```

Comment:
When an UnsupportedEncodingException is thrown, is url null and does it then throw an exception here? Maybe place the return statement inside the try {} block in order to prevent this from happening.